File: csrc/elastic_solid/mpm_solve.c
//Standard Includes
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h> // for benchmarking

// Constants and options
#include "solver_options.h"

//Custom Built Functions that contain necessary definitions
#include "mpm_math.h"

//Class definitions that define the mesh and the material points
#include "node.h"
#include "material_point.h"

//Information about the initial State
#include "initial_state.h"

//Information about the external accelerations of the system
#include "accelerations.h"

//Functions that let us initialize the mesh and material points
#include "initializations.h"

//Functions for advancing the system in time
#include "sim_update.h"//For density, etc

//Data Exportation and Utility
#include "data_write.h"//Functions on writing data to a CSV



int main() {


// Time
	double time_spent = 0.;
	clock_t begin = clock();

// Initial simulation setup

	// The time
	double t = 0.;

	//Generic elements of classes
	struct node nodes[num_nodes];
	struct material_point mp_points[num_particles];
	
	

//------------------------------------------------------------------------------------
// Data Setup



	// Create output data files
	FILE *nodeData;
	FILE *mpData;
	FILE *debugFile;

	nodeData = fopen( node_filename, "w+" );
	mpData = fopen( mp_filename, "w+" );


	// Write header for the output files
	write_header( nodeData, node_titles );
	write_header( mpData, mp_titles );



	
	
//--------------------------------------------
//INITIALIZATION


	// First initialize the mesh
	// Then use the mesh to initialze the material points
	//	In the material points initializiation, we find the nearest nodes
	//	to each particles, which we fix for the whole program.
	//		See: initializations.c
	//			 initial_state.c
	initialize_mesh(nodes);
	initialize_material_points(mp_points, nodes);
	

	// Using the material points to approximate an integral to compute the mass, 'inverse mass'
	// and Y equivalent for use in time derivatives
	//		See sim_update.c
	compute_node_masses(nodes, mp_points);





//TIME EVOLUTION
		
	for (int tt = 0; tt < tsteps; ++tt) {


		if (tt % record_frequency == 0) {

			// Write the data
			node_data_write( nodeData, nodes, t );
			mp_data_write( mpData, mp_points, t );

		}//end if
		
		
		// Then evolve the system
		//
		//
		// First update the strain in time
		//
		// Then the velocity:
		// Given the velocity at one moment of time, advances it to the next timestep
		// Does:
		//	1. Partially calculate the mp velocity at the next time
		//	2. Calculate the lagr xvel of a node
		//	3. Calulate the physical velocity of the particles
		//	4. Finishes calculating the mp velocity at the next time
		//	5. (optional) recomupute particle positions + node masses
		//	6. Update the node velocity using the new mp velocities
		//
		// Then update the stress
		update_particle('e', mp_points, nodes);
		update_velocity( nodes, mp_points, t );
		update_particle('s', mp_points, nodes);
		



		
		//Update the other mesh quantities
		// See: sim_update.c
		//	Denisty
		//	Strain
		//	Stress
		update_node('r', nodes, mp_points);
		update_node('e', nodes, mp_points);
		update_node('s', nodes, mp_points);


		// Advance the system in time
		t += dt;
	
		
	}//end for
	
//--------------------------------------------------------


//-------------------------------------------------------
//Finshing up writing data


	node_data_write( nodeData, nodes, t );
	mp_data_write( mpData, mp_points, t );
	fclose(nodeData);
	fclose(mpData);
	


	// Print the time elapsed
	clock_t end = clock();

	time_spent += (double)(end-begin) / CLOCKS_PER_SEC;

	printf("Time elapsed: %f seconds\n", time_spent);

	return 0;


}//end main

File: javasrc/elastic_solid/MPMSolve.java
import java.io.*;
import java.util.*; // Get lists, etc


// Our main program
// Puts everything together
public class MPMSolve {



	public static void main( String args[] ) throws IOException {


		// Get start time
		final long startTime = System.currentTimeMillis();


//---------------------------------------------------------------
		// SETUP



		// Our time
		double t = 0.;




		// Our data files
		FileWriter nodeData = new FileWriter(Constants.node_csv_name);
		FileWriter mpData = new FileWriter(Constants.mp_csv_name);
		FileWriter debugData = new FileWriter("debug.txt");

		if ( !Constants.debug_file ) {

			debugData.append("No debug file produced.\n");
			debugData.flush();
			debugData.close();		

		}//end if


		


		// Give our data files some headers!
		DataWrite.MakeHeaders(nodeData, Constants.node_titles);
		DataWrite.MakeHeaders(mpData, Constants.mp_titles);

		if ( Constants.debug_file ) {

			Debug.DataDumpHeader(debugData);

		}//end if



//-----------------------------------------------------------------
		// Creating data arrays



		// Our current-time data
		List<Node> nodes = new ArrayList<Node>();
		List<MaterialPoint> mp_points = new ArrayList<MaterialPoint>();



		// Populate our lists with enough entries to make an initial state
		for ( int i = 0; i < Constants.num_nodes; ++i ) {


			nodes.add( new Node() );


		}//end for

		for ( int i = 0; i < Constants.num_particles; ++i ) {


			mp_points.add( new MaterialPoint() );


		}//end for




//-----------------------------------------------------------------
		// Initializing our system		




		// Then make the initial state
		Initializations.InitializeMesh(nodes);
		Initializations.InitializeMaterialPoints(mp_points, nodes);
		


		// Then get the nearest nodes to our material points
		// and find the initial node masses
		MPMMath.GetNearNodes(mp_points);
		SimUpdate.ComputeNodeMasses(nodes, mp_points);




//-----------------------------------------------------------------
		// Time evolution




		// Then evolve the system in time
		for ( int tt = 0; tt < Constants.tsteps; ++tt ) {



			// Data recording
			if (tt % Constants.record_frequency == 0) {

				// First save our data
				DataWrite.Node(nodeData, nodes, t);
				DataWrite.MaterialPoint(mpData, mp_points, t);

				if ( Constants.debug_file ) {

					Debug.DataDump(debugData, nodes, mp_points, t);

				}//end if

			}//end if


			// Then use these Lagrangian quantities to find the Eulerian ones
			// At the next timestep
			SimUpdate.UpdateParticle.Strain(mp_points, nodes);
			SimUpdate.UpdateVelocity(nodes, mp_points, t);
			SimUpdate.UpdateParticle.Stress(mp_points, nodes);



			// Then update the node quantities
			SimUpdate.UpdateNode.Density(nodes, mp_points);
			SimUpdate.UpdateNode.Strain(nodes, mp_points);
			SimUpdate.UpdateNode.Stress(nodes, mp_points);



			// Advance system in time
			t += Constants.dt;




		}//end for


		// Final save of the data
		DataWrite.Node(nodeData, nodes, t);
		DataWrite.MaterialPoint(mpData, mp_points, t);




		// Flush and close our files
		nodeData.flush();
		mpData.flush();

		nodeData.close();
		mpData.close();


		if ( Constants.debug_file )  {

			debugData.flush();
			debugData.close();

		}//end if
		

		// Get final time and print total time taken to run the program
		final long endTime = System.currentTimeMillis();

		System.out.println("Total execution time: " + (endTime - startTime));


	}//end main
	




}//end MPMSolve
